### PR TITLE
Input to cadence with type hint

### LINKED ISSUE #10: Parse input to cadence with typehint
Add a new method that allows parsing from any input to cadence.Value with an typehint.

### PR DESCRIPTION
**Pull request by bjartek**
closes #10

### CODE REVIEW

File: input.go [hunk 1]
```diff
@@ -7,10 +7,200 @@ import (
...
+}
+
 // a resolver to resolve a input type into a name, can be used to resolve struct names for instance
-type InputResolver func(string) (string, error)
+type InputResolver func(string, ResolveType) (string, error)
+
+var flowInterpeter, _ = interpreter.NewInterpreter(nil, nil, &interpreter.Config{})
+
+func InputToCadenceWithHint(v interface{}, typeHint sema.Type, resolver InputResolver) (cadence.Value, error) {
+	// if we are already a cadence value then we just return
```
Comment (willsu88) [inline input.go]:
```suggestion
	
```
nit pick: i think the if statement at line 35 makes this p clear already, we could opt to discard the comment

File: input.go [hunk 2]
```diff
@@ -7,10 +7,200 @@ import (
...
+			}
+			cadenceAddress := cadence.BytesToAddress(adr.Bytes())
+			return cadenceAddress, nil
+		} else if th == sema.StringType {
+			if len(stringVal) > 0 && !strings.HasPrefix(stringVal, "\"") {
+				stringVal = "\"" + stringVal + "\""
+			}
+		}
+
+		// we need to check if this is an optional string
```
Comment (willsu88) [inline input.go]: just curious, is there a reason why we aren't doing this in this PR?

File: input.go [hunk 3]
```diff
@@ -7,10 +7,200 @@ import (
...
+		return cadence.NewBool(value.Interface().(bool)), nil
+	case reflect.String:
+		stringVal := value.Interface().(string)
+		th := typeHint
+		optionalHint, isOptionalType := typeHint.(*sema.OptionalType)
+		if isOptionalType {
+			th = optionalHint.Type
+		}
+		if th == sema.TheAddressType {
+			// TODO: do we need to have different types of resolvers here?
```
Comment (willsu88) [inline input.go]: might be good to add this to our backlog story

File: input.go [hunk 4]
```diff
@@ -7,10 +7,200 @@ import (
...
+			QualifiedIdentifier: resolvedIdentifier,
+			Fields:              fields,
+		}
+
+		structValue := cadence.NewStruct(val).WithType(&structType)
+		return structValue, nil
+
+	case reflect.Pointer:
+		/*
+			*  //not sure when this will fire?
```
Comment (willsu88) [inline input.go]: is this testable currently? do we need to track this in a story?

File: input.go (hunk 3 above)
Comment (bjartek) [inline input.go]:
```suggestion
```

Comment (bjartek) [reply input.go]: i have removed this as we do have types for resolvers now.

File: input.go (hunk 2 above)
Comment (bjartek) [inline input.go]:
```suggestion
```

Comment (bjartek) [reply input.go]: this comment is also dead, it is handled.

File: input.go (hunk 4 above)
Comment (bjartek) [reply input.go]: This code is also dead, we do not need it, i belive there are tests that test that if we have a pointer in we will accept both an optional type hint or just a type hint that is a value.

File: input_test.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+			input:    "first",
+			typeHint: sema.TheAddressType,
+		},
+		{
+			want:     autogold.Want("stringPointer", cadence.Optional{Value: cadence.String("foobar")}),
+			input:    &stringVal,
+			typeHint: &sema.OptionalType{Type: sema.StringType},
+		},
+		{
+			want:     autogold.Want("string pointer empty", cadence.Optional{Value: cadence.String("foobar")}),
```
Comment (bjartek) [inline input_test.go]: this test must be wrong, it makes no sense it is basically the same as the one above is it not?

File: input_test.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.want.Name(), func(t *testing.T) {
+			val, err := InputToCadenceWithHint(tc.input, tc.typeHint, resolver)
+			require.NoError(t, err)
+			tc.want.Equal(t, val)
+		})
+	}
+	/*
```
Comment (bjartek) [inline input_test.go]:
```suggestion
```

File: input_test.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+				vj, err := json.Marshal(v)
+				assert.NoError(t, err)
+
+				cvj, err := json.Marshal(value)
+				assert.NoError(t, err)
+
+				assert.Equal(t, string(cvj), string(vj))
+			})
+		}
+	*/
```
Comment (bjartek) [inline input_test.go]:
```suggestion
```

File: input_test.go (hunk 5 above)
Comment (willsu88) [reply input_test.go]:
```suggestion
			want:     autogold.Want("string pointer empty", cadence.Optional{Value: cadence.String("")}),
```
we want it to be empty string? or should the optional just be completely empty (eg without even the cadence.String value)?

Comment (willsu88) [reply input_test.go]: maybe` cadence.NewOptional(nil)` . not sure with the syntax

Comment (bjartek) [reply input_test.go]: i removed this test in my latest commit, but not sure if we should add one with empty string.

Comment (willsu88) [reply input_test.go]: maybe we can try to keep it if we can, no harm there

**Review by willsu88 — APPROVED** [review]
LGTM. I think just need to run `go mod tidy ` so that the github pipeline passes

**Pull request merged**